Add AureliaSpa template

kmkatsma opened the pull request:
This PR adds a new AureliaSpa folder to the existing templates folder.

Made attempts to be as consistent as possible with existing templates, but this does follow different webpack template build process - it uses a single webpack config utilizing the AureliaWebpackPlugin.  However, I tried to keep the WebPack configuration consistent with the configurations that exist in official Aurelia webpack skeletons, except EasyWebpack has been removed for closer match to the existing webpack templates in this repo.

Other things of note with regard to comparison with other templates:
1. The nav css had to be modified to deal with the default class added by aurelia to elements (perhaps there is another way).
2. The menu is build dynamically, instead of statically
3. Single boot file, given no server side rendering
4. The fetchdata.ts has some typings overrides added specifically to deal with some current issues in "official" typings files

@SteveSandersonMS, @MarkPieszak, @EisenbergEffect

dnfclas commented:
Hi __@kmkatsma__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla2.dotnetfoundation.org.
    </span>

TTYL, DNFBOT;

dnfclas commented:
__@kmkatsma__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, DNFBOT;

SteveSandersonMS commented:
Thanks @kmkatsma! Overall this looks great. You're identified most of the limitations yourself above, and I appreciate that they are beyond your control. Thanks for making it consistent with the other templates where possible.

I'll now review and comment on the actual code. Since I don't know Aurelia, I'll wait for @EisenbergEffect to comment on whether Aurelia is being used optimally. Mostly I'll be limited to commenting on the cosmetic aspects of the code - hope you don't mind my rather trivial comments!

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/app/app.html:
```diff
@@ -0,0 +1,14 @@
+<template>
```
Inconsistent quote style in this file. In TypeScript let's use single-quotes. In HTML templates, I don't know if Aurelia community has a standard, but can you find out what it is and use that?

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/app/app.html:
```diff
@@ -0,0 +1,14 @@
+            </div>
+        </div>
+    </div>
+</template>
```
Missing trailing linebreak which is present in most files. Could you add one here and to any other TS/JS/HTML/CSS file where it's missing?

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/app/app.ts:
```diff
@@ -0,0 +1,19 @@
+import {Aurelia} from 'aurelia-framework';
```
Some few formatting issues in this file:
 * missing spaces around `{ ImportedModule }`
 * The `config.map` object looks pretty strange. Not sure if you were trying to line things up with tabs or something, but it's not lined up now. Also some of the key-value pairs have spaces after the colon and others don't. Could you lay this code out more conventionally?

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/app/app.ts:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+ 
```
Line 18 seems like unnecessary whitespace, then there's a missing trailing linebreak.

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/app/app.ts:
```diff
@@ -0,0 +1,19 @@
+  configureRouter(config: RouterConfiguration, router: Router) {
+    config.title = 'Aurelia';
+    config.map([
+      { route: ['', 'home'], name: 'home', settings:{class:'home'},     moduleId: '../home/home',  nav: true, title: 'Home' },
```
The use of the term `class` here is a bit odd. I'd naturally think that refers to a class in the code, but it looks like you're using that to mean "CSS class we should set on the icon element". Could you rename this to something more descriptive, such as `icon` or `icon-className`?

EisenbergEffect replied:
Prefer `icon`.

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/counter/counter.ts:
```diff
@@ -0,0 +1,9 @@
+
```
Unnecessary blank lines on line 1 and 7

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/fetchdata/fetchdata.html:
```diff
@@ -0,0 +1,26 @@
+<template>
```
Inconsistent quotes and `${ spacing }` in this file

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/fetchdata/fetchdata.ts:
```diff
@@ -0,0 +1,27 @@
+/// <reference path="../../../../node_modules/aurelia-fetch-client/doc/whatwg-fetch.d.ts" />
+/// <reference path="../../../../node_modules/aurelia-fetch-client/doc/url.d.ts" />
```
It's unfortunate that these are necessary. Has the underlying issue in Aurelia been reported? Is there any chance of getting a fix published?

EisenbergEffect replied:
These don't have anything to do with Aurelia as far as I know. They are standards that TypeScript is missing for some reason. However, I don't think that `reference` comments should be needed either way.

SteveSandersonMS replied:
@EisenbergEffect I think @kmkatsma was saying they are needed because `aurelia-fetch-client` currently contains an incorrect `.d.ts` file.

EisenbergEffect replied:
As far as we know, there's nothing incorrect in our dts. It's just that we use the Fetch spec which relies on the there being dts for fetch and url. Those should be provided by TS because they are standards, but they aren't, so you have to include your own. If there's something we need to fix on our side, we're happy to do it though.

kmkatsma replied:
Yes, in general the change in type references for TS 2.0 was something I only realized this Saturday as I was cleaning up some things.  Will look at.

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/fetchdata/fetchdata.ts:
```diff
@@ -0,0 +1,27 @@
+@inject(HttpClient)
+export class Fetchdata {
+    public forecasts: WeatherForecast[];
+    http: HttpClient;
```
Why store this as a member property? It doesn't seem to be used anywhere outside the constructor.

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/home/home.html:
```diff
@@ -0,0 +1,18 @@
+<template>
```
This file is about the Angular 2 template. Could you update it?

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/app/components/navmenu/navmenu.html:
```diff
@@ -0,0 +1,26 @@
+<template bindable="router">
```
Inconsistent quotes in this file

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/boot.ts:
```diff
@@ -0,0 +1,15 @@
+import 'bootstrap/dist/css/bootstrap.css';
+import 'bootstrap';
+ 
+// comment out if you don't want a Promise polyfill (remove also from webpack.config.js)
```
TBH I'd prefer not to include bluebird in the template, just because the template would be a little simpler without it. People can add their own choice of promise library if they want.

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/boot.ts:
```diff
@@ -0,0 +1,15 @@
+import 'bootstrap';
+ 
+// comment out if you don't want a Promise polyfill (remove also from webpack.config.js)
+import * as Bluebird from  'bluebird';
```
Layout issues in this file (e.g., double-whitespace on line 1, uses two-char indentation elsewhere when we use four-char in all the other templates).

SteveSandersonMS commented on templates/AureliaSpa/ClientApp/boot.ts:
```diff
@@ -0,0 +1,15 @@
+export function configure(aurelia: Aurelia) {
+  aurelia.use
+    .standardConfiguration()
+    .developmentLogging();
```
Can you adapt this so that the `developmentLogging` only happens from dev builds and not from prod builds?

kmkatsma replied:
Looks like I'll add a check for localhost in window location to decide whether to enable logging.  Couldn't find a better suggestion not relying on a gulp task to actually edit code (that's what I've used in past).  Only happens when app is loading so doesn't seem too bad.

SteveSandersonMS commented on templates/AureliaSpa/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM microsoft/dotnet:latest
```
This file is slightly out of date (compared with the other templates). Would you be able to update it, e.g., based on https://github.com/aspnet/JavaScriptServices/blob/dev/templates/Angular2Spa/Dockerfile?

SteveSandersonMS commented on templates/AureliaSpa/README.md:
```diff
@@ -0,0 +1,40 @@
+# Welcome to ASP.NET 5
```
This file is out of date. Could you update it? e.g., take content from https://github.com/aspnet/JavaScriptServices/blob/dev/templates/Angular2Spa/README.md

EisenbergEffect commented on templates/AureliaSpa/ClientApp/app/components/navmenu/navmenu.html:
```diff
@@ -0,0 +1,26 @@
+            <div class='clearfix'></div>
+            <div class='navbar-collapse collapse'>
+                <ul class='nav navbar-nav'>
+                    <li class="${row.isActive ? 'link-active' : ''}" repeat.for = "row of router.navigation" >
```
For clarity, I would put the repeat.for first on the `li`, before the `class` binding.

SteveSandersonMS commented on templates/AureliaSpa/Views/Home/Index.cshtml:
```diff
@@ -0,0 +1,9 @@
+    ViewData["Title"] = "Home Page";
+}
+ 
+@section scripts {
```
Any reason why the client should be doing 3 separate HTTP requests for these JS files? Couldn't the `bootstrap` and `modules` bundles be a single bundle? (Ideally, all the vendor dependencies would be in a single file)

SteveSandersonMS commented on templates/AureliaSpa/Views/Shared/_Layout.cshtml:
```diff
@@ -0,0 +1,16 @@
+        <title>@ViewData["Title"] - Aurelia</title>
+        <base href="/" />
+    </head>
+    <body aurelia-app="boot">
```
Is it definitely correct to have an `aurelia-app` attribute? There's a discussion at https://github.com/aurelia/bootstrapper-webpack/issues/2 that implies (for reasons I don't follow) that when you're using Webpack, you don't need `aurelia-app`. @EisenbergEffect, could you comment on whether this app is bootstrapping itself in the desired way?

SteveSandersonMS replied:
Even if this is needed, could you move it into `Views/Home/Index.cshtml`, since it's only that one view that really needs to start up the Aurelia app?

SteveSandersonMS commented on templates/AureliaSpa/package.json:
```diff
@@ -0,0 +1,52 @@
+    "aurelia-templating-binding": "^1.0.0",
+    "aurelia-templating-resources": "^1.0.0",
+    "aurelia-templating-router": "^1.0.0",
+    "bluebird": "^3.4.1",
```
Ideally, remove this and `@types/bluebird`

SteveSandersonMS commented on templates/AureliaSpa/package.json:
```diff
@@ -0,0 +1,52 @@
+    "bootstrap": "^3.3.7",
+    "isomorphic-fetch": "^2.2.1",
+    "jquery": "^2.2.1",
+    "preboot": "^4.5.2"
```
Are you really using this `preboot` library? Can you check and remove any other NPM dependencies that aren't actually needed?

SteveSandersonMS commented on templates/AureliaSpa/package.json:
```diff
@@ -0,0 +1,52 @@
+    "jquery": "^2.2.1",
+    "preboot": "^4.5.2"
+  },
+  "devDependencies": {
```
I'm not sure it's valid to regard these as `devDependencies`. When you're publishing to production, isn't the publish script set up so that it would only restore `dependencies`, and then the webpack compilation would fail?

SteveSandersonMS commented on templates/AureliaSpa/package.json:
```diff
@@ -0,0 +1,52 @@
+    "html-loader": "^0.4.4",
+    "html-webpack-plugin": "^2.22.0",
+    "raw-loader": "^0.5.1",
+    "sourcemap-istanbul-instrumenter-loader": "^0.2.0",
```
What's this for?

SteveSandersonMS commented on templates/AureliaSpa/project.json:
```diff
@@ -0,0 +1,76 @@
+      "tsconfig.json",
+      "tsd.json",
+      "web.config",
+      "webpack.*.js",
```
Why would that be published?

EisenbergEffect commented:
I left one minor style comment related to the Aurelia code. Everything else looks fine to me :)

SteveSandersonMS commented on templates/AureliaSpa/project.json:
```diff
@@ -0,0 +1,76 @@
+      "ClientApp/dist",
+      "node_modules",
+      "tsconfig.json",
+      "tsd.json",
```
I don't think `tsd.json` even exists, and I don't think `tsconfig.json` would need to be published. Can you check that the publish mechanism works and only publishes necessary things?

SteveSandersonMS commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+var path = require('path');
+var webpack = require('webpack');
+var AureliaWebpackPlugin = require('aurelia-webpack-plugin');
+const srcDir = path.resolve('./ClientApp');
```
Not sure why some of these are `const` and others are `var`. Ideally we'd support Node 0.x which is ES5. If we strictly need ES2015, can we be consistent about using `const`/`let`/etc?

SteveSandersonMS commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+     entry: {
+        'app': [], // <-- this array will be filled by the aurelia-webpack-plugin
+        'aurelia-bootstrap': aureliaBootstrap,
+        'aurelia-modules': aureliaModules.filter(pkg => aureliaBootstrap.indexOf(pkg) === -1)
```
Why are these dependencies going into a separate bundle? Couldn't they just be included in `aurelia-bootstrap`? (which they would be by default if this line wasn't here)

SteveSandersonMS commented on templates/AureliaSpa/wwwroot/web.config:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
I don't think this file is needed. It's not present in any of the other templates.

MarkPieszak commented:
Forgot to say, great job @kmkatsma ! :100: :+1:

SteveSandersonMS commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+    },
+    plugins: [
+        new webpack.ProvidePlugin({
+            regeneratorRuntime: 'regenerator-runtime', // to support await/async syntax
```
Is that needed? I didn't see any `await`/`async` anywhere? Can this be simplified?

SteveSandersonMS commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,74 @@
+            Promise: 'bluebird', // because Edge browser has slow native Promise object
+            $: 'jquery', // because 'bootstrap' by Twitter depends on this
+            jQuery: 'jquery', // just an alias
+            'window.jQuery': 'jquery' // this doesn't expose jQuery property for window, but exposes it to every module
```
Do we need all three of these aliases for jQuery?

EisenbergEffect commented:
Agree with @MarkPieszak This is a really awesome contribution. When it this goes live, if you would like to @kmkatsma we can feature you on our official blog for this work. I'm sure a lot of people will be interested.

SteveSandersonMS commented on templates/AureliaSpa/Startup.cs:
```diff
@@ -0,0 +1,67 @@
+            {
+                app.UseDeveloperExceptionPage();
+                app.UseWebpackDevMiddleware(new WebpackDevMiddlewareOptions {
+                    HotModuleReplacement = false
```
I'm not sure we'll get away with just setting this to `false`. The problem is, people will change it to `true`, and then you get a really weird and non-descriptive error that people will regard as a bug. It's reasonable for people to think that if some option exists that you should be able to use it.

Obviously the ideal solution would be if HMR could be made to work. I'd hope that's not impossible, since it's quite trivial to enable with Angular/Knockout/React, but I don't know if Aurelia does something that can't handle it.

If it's not possible to make work, can you somehow stop the unrelated-seeming error message, and replace it with something like "Sorry, Hot Module Replacement is not supported with Aurelia"?

kmkatsma commented:
@SteveSandersonMS - thanks - a lot of good comments here, responded to couple, but too many to address individually. =)  I will look at them all.  I was trying to clean things up for consistency, and this helps a ton.

@MarkPieszak @EisenbergEffect - thanks, appreciate it!  I would definitely be good with a feature on the Aurelia blog!  These kinds of projects are really important - anything we can do to help developers get a foothold into technologies like these is worthwhile.  And I think Aurelia deserved a place here as well. =)

EisenbergEffect commented:
I think doing this sort of thing in a constructor is bad, yes. I wouldn't
recommend that in any language. Constructor code should always by
synchronous.

On Mon, Oct 24, 2016 at 6:54 AM, Steve Sanderson <notifications@github.com>
wrote:

> *@SteveSandersonMS* commented on this pull request.
> ------------------------------
>
> In templates/AureliaSpa/ClientApp/app/components/fetchdata/fetchdata.ts
> <https://github.com/aspnet/JavaScriptServices/pull/398>:
>
> > @@ -0,0 +1,27 @@
> +/// <reference path="../../../../node_modules/aurelia-fetch-client/doc/whatwg-fetch.d.ts" />
> +/// <reference path="../../../../node_modules/aurelia-fetch-client/doc/url.d.ts" />
> +import {HttpClient} from 'aurelia-fetch-client';
> +import {inject} from 'aurelia-framework';
> +
> +@inject(HttpClient)
> +export class Fetchdata {
> +    public forecasts: WeatherForecast[];
> +    http: HttpClient;
>
> @EisenbergEffect <https://github.com/EisenbergEffect> It's the
> simplicity/extensibility tradeoff. In the Angular/Knockout/React templates
> we've followed this exact approach, as we're trying to make the templates
> accessible to newcompers and limit the concept count. People who get into
> using one of those frameworks won't struggle to create a service layer. In
> the ReactRedux template, we do the whole Redux thing which is a whole extra
> level of sophistication, but that's just fundamental to doing Redux. I
> think for this Aurelia template it's a reasonable choice of simplicity vs
> extensibility already, but we could change it if you think it's just flat
> out bad.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/aspnet/JavaScriptServices/pull/398>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAIBnccSp7t35BhGwypRsjQjHbEvaTTAks5q3LiwgaJpZM4Kd5Pj>
> .
>



-- 
Rob Eisenberg,
President - Blue Spire
aurelia.io

SteveSandersonMS commented:
@EisenbergEffect It's a reasonable point but perhaps we're getting a bit side-tracked.

@kmkatsma's Thanks - awaiting your updates :)

kmkatsma commented:
Have pushed in update in updates for all the review items.  There's two things to note regarding resolutions (or somewhat lack thereof ).

1.  For fetchdata, I couldn't eliminate need for the d.ts custom entries using the new types approach.  I could move to a typings folder, but I think that is not any better.  I also left the api call in constructor for now since it's same as Angular2 template, but would be willing to refactor this.

1.  For dotnet publish, it builds, and I have its output definition matching angular2 template, but I am having issues getting the publish of either Angular2 or Aurelia running in IIS on my dev VM (built off the new windows 10 pro dev images).  I spent a fair amount of time on this, working through some stackoverflow answers to no avail.  I can spend some more time if we want to get a running confirmation before merging.

2. HMR - I was able to get app to run with this setting enabled using a hack to webpack config, but it caused issues when running in asp env production mode.   So at this point, I put a comment in Startup.cs to the effect that HMR has issues, and to leave the setting false.  Definitely this is something that is high priority for developer experience, but the webpack integration with Aurelia is still work in progress.   I think the convention based approach (which is advantage from code perspective) actually makes it hard for WebPack to find all its pieces, and so are dependent on the custom plugin to pull in all the pieces and run it.

Let me know thoughts on these things - have learned a ton here going through this!  I think the back end integration of these templates is a *huge* win, and removes a lot of friction we've seen since everything moved to client in a SPA.  I can definitely see people wanting to get more things into them.  =)

kmkatsma commented:
Added a commit that fixed issues with dotnet publish.  

First, when publishing to virtual directory that is not at root, the router was having issues due to base url set to "\" in the layout.cshtml.  Apparently this works for angular2 in publish mode, but causes issues for aurelia routing.  I removed this, and now works in dev and publish mode under IIS virtual directory.

Second change was to remove node_modules from the items included in publish process.  It made the publish really big and run really slow, and is not needed since webpack runs as pre-publish step.

There's a funky bug in publish process where the generated web.config keeps appending extra arguments on each run if you don't delete the web.config first.  This was causing my publish problems.

niieani commented on templates/AureliaSpa/package.json:
```diff
@@ -0,0 +1,43 @@
+    "jquery": "^2.2.1"
+  },
+  "devDependencies": {
+    "@types/node": "^6.0.45",
```
What are the `node` types used for?

niieani commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,64 @@
+     resolve: { extensions: [ '.js', '.ts' ] },
+     devtool: isDevBuild ? 'inline-source-map' : null,
+     entry: {
+        'app': [], // <-- this array will be filled by the aurelia-webpack-plugin
```
Latest Webpack@beta currently errors when an array is empty (it's an ongoing bug). I might be better to instead point it to where the `main.js` file is, as that should be in the context anyway.

SteveSandersonMS replied:
This is changed in the current version too, so I think we're OK.

niieani commented on templates/AureliaSpa/webpack.config.js:
```diff
@@ -0,0 +1,64 @@
+var outDir = path.resolve('./wwwroot/dist');
+var baseUrl = '/';
+var project = require('./package.json');
+var aureliaModules = Object.keys(project.dependencies).filter(dep => dep.startsWith('aurelia-'));
```
This only gets the list of first-level dependencies of Aurelia, which means not all of Aurelia's dependencies might get into the `aurelia-modules` bundle specified in the `entry` config.

SteveSandersonMS replied:
Good point. However this approach is no longer used in the current version anyway - dependencies are now listed explicitly.

SteveSandersonMS commented:
Thanks again @kmkatsma - this is now merged :) Also I've published `generator-aspnetcore-spa` version 0.4.0 in which you'll find the new "Aurelia" option.

I made a few further tweaks - hope you don't mind:

 * Switched to use the same 'vendor' bundle pattern as the other templates
 * Switched to use external source maps configured for VS/VSCode compatibility (again, to match the other templates)
 * Changed the "log or not" choice to depend on which webpack build configuration you used, rather than based on whether you have `localhost` in your hostname
 * Minor stylistic tweaks

MarkPieszak commented:
@niieani The `@types/Node` is mainly to make Visual Studio happy when it comes to `require` and other similar things.

EisenbergEffect commented:
Awesome! @kmkatsma We'd love to have you featured on the official Aurelia blog talking about this new options for our customers :) Please email me and let's make some plans. Thanks for the great work!

ClaudioNunes commented:
The only thing missing is HMR. Maybe @niieani could help with this.

niieani commented:
HMR support is planned. Once major Aurelia+Webpack fixes land in November, I might tackle basic HMR in December. We'll see how difficult it will be.

kmkatsma commented:
@SteveSandersonMS - your changes wrapped this up nicely, I think!

ClaudioNunes commented:
@niieani  any news about HMR support?

niieani commented:
@ClaudioNunes not yet. But work is on-going. See https://github.com/aurelia/skeleton-navigation/pull/714 for a couple of updates (HMR for CSS works).
